File: 백준/Gold/1753. 최단경로/최단경로.py
'''
Condition
    - TL: 1s
    - 

Goals
    - 시작점에서 다른 모든 정점으로의 최단 경로의 경로값을 출력

Approach
    - dijkstra
    - TC = O((V+E)logV) = O(약 32만)
'''
from sys import stdin
from heapq import *
input = stdin.readline

def input_data():
    v,e = map(int,input().split())
    s = int(input())
    graph = [[] for _ in range(v+1)]
    for _ in range(e):
        stt,end,cost = map(int,input().split())
        graph[stt].append((end,cost))
    return v,e,s,graph

def dijkstra(s:int, graph:list[list]):
    global distance
    hq = []
    heappush(hq, (0, s))
    while hq:
        dist, node = heappop(hq)
        if distance[node] < dist:
            continue
        # node의 인접 노드를 탐색
        for data in graph[node]:
            neighbor = data[0]
            cost = distance[node] + data[1]
            if cost < distance[neighbor]:
                distance[neighbor] = cost
                heappush(hq, (cost, neighbor))
    return

def main(v:int, e:int, s:int, graph:list[list]):
    global distance
    INF = int(1e9)
    distance = [INF] * (v+1)
    distance[s] = 0
    dijkstra(s, graph)
    # 출력
    for i in range(1,v+1):
        print(distance[i] if distance[i]!=INF else "INF")
    return

main(*input_data())

File: 백준/Gold/1753. 최단경로/최단경로.java

import java.io.*;
import java.util.*;


public class Main {

	static StringTokenizer st;
	static StringBuilder sb = new StringBuilder();
	static int v;
	static int e;
	static int k;
	static int x;
	static int[] dist;
	static List<Node>[] graph;
	
	
	public static void main(String[] args) throws IOException {

		// input
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));		
		
		st = new StringTokenizer(br.readLine());
		v = Integer.parseInt(st.nextToken());
		e = Integer.parseInt(st.nextToken());
		
		k = Integer.parseInt(br.readLine());
		
		dist = new int[v+1];
		for (int i=1; i<v+1; i++) dist[i] = Integer.MAX_VALUE;
		
		graph = new ArrayList[v+1];
		for (int i=0; i<v+1; i++) graph[i] = new ArrayList<>();
		for (int i=0; i<e; i++) {
			
			st = new StringTokenizer(br.readLine());
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());
			int w = Integer.parseInt(st.nextToken());
			
			graph[u].add(new Node(v, w));
		}
		
		// dijkstra
		dijkstra();
		
		// result
		for (int i=1; i<v+1; i++) {
			sb.append(dist[i] != Integer.MAX_VALUE ? dist[i] : "INF").append("\n");
		}
		System.out.print(sb);
		
	}
	
	
	
	static void dijkstra() {
		
		// 초기화
		Queue<Node> q = new PriorityQueue<Node>((o1, o2) -> o1.w - o2.w);
		q.offer(new Node(k, 0));
		dist[k] = 0;
		
		while (!q.isEmpty()) {
			
			Node curNode = q.poll();
			int curIdx = curNode.idx;
			int curW = curNode.w;
			
			// cur을 지나는게, 현재 등록된 dist보다 비싸면 pass
			if (dist[curIdx] < curW) continue;
			
			// 탐색
			for (int i=0; i<graph[curIdx].size(); i++) {
				
				Node next = graph[curIdx].get(i);
				int nextIdx = next.idx;
				int cost = curW + next.w;
				
				if (cost < dist[nextIdx]) {
					
					dist[nextIdx] = cost;
					q.offer(new Node(nextIdx, cost));
					
				}
				
			}	
		}	
	}
	
	
	
	static class Node {
		
		int idx;
		int w;
		
		Node(int idx, int w) {
			this.idx = idx;
			this.w = w;
		}
		
	}
	
	
	
}